runtime/include/chpl-comm-callbacks.h

```diff
@@ -0,0 +1,179 @@
+//   is of interest. The 'cb_fn' argument is a pointer to a callback
+//   function for that event.  The callback functions installed for
+//   each event will be called, in the order in which they were
+//   installed, each time the corresponding event occurs in the
+//   communication layer of the runtime.
+//
+//   The event type, chpl_comm_cb_event_kind_t is defined below.
+//
+//   The type of a communication layer callback function pointer is:
+//
+//     typedef void (*chpl_comm_cb_fn_t)(chpl_comm_cb_info_t* info);
+//
+//   The single argument passed to it is a pointer to information of
+//   type chpl_comm_cb_info_t, describing the task and/or event.  This
+//   type is also defined below.  It contains a union of three different
+//   information sets dependinf only on the chpl_comm_cb_devent_kind_t
+//   value.   
+//
+//
+// RETURN VALUE
+//
+//   The returned value is 0 if no errors occurred and some other value
+//   (with errno set) if any errors did occur.
+//
+//
+// ERRORS
+//
+//   The following errors can occur with chpl_comm_install_callback():
+//
+//     ENOMEM:  No room to install another callback function for this
+//              event.  At present there is a static limit of 10 installed
+//              callback functions for each event.
+//     ERANGE:  The specified 'kind' is too large.
+//
+//   The following errors can occur with chpl_comm_uninstall_callback():
+//
+//     ENOENT:  The given pointer was not found in the list of installed
+//              callback functions for the given event.
+//     ERANGE:  The specified 'kind' is too large.
+//
+//
+// NOTES
+//
+//   Communication callbacks are local to the top-level locale.  In 
+//   multi-locale programs, callback functions must be installed separately
+//   on each locale where they are wanted.
+//
+//   The communication callback system is not thread safe.  Calling the install
+//   and/or uninstall functions simultaneously from more than one thread
+//   on a single locale can corrupt internal data structures and lead to
+//   chaos.  The installed callback functions themselves may be called
+//   by more than one thread on the same locale simultaneously; if they
+//   need concurrency control they must provide for it themselves.
+//
+//   The callback function should not depend upon the allocation
+//   status of the pointed-to info or any member within it, past the
+//   point when it returns.  For example, it must not store a copy of
+//   the info pointer for an event call for use when any other event
+//   call occurs.
+//
+
+typedef enum {
+  chpl_comm_cb_event_kind_put,       // Simple put
+  chpl_comm_cb_event_kind_put_nb,    // Non-blocking put
+  chpl_comm_cb_event_kind_put_strd,  // Strided put
+  chpl_comm_cb_event_kind_get,       // Simple get
+  chpl_comm_cb_event_kind_get_nb,    // Non-blocking get
+  chpl_comm_cb_event_kind_get_strd,  // Strided get
+  chpl_comm_cb_event_kind_fork,      // regular fork
```

Comment:
Let's use "executeOn" instead of "fork".  The latter is the old terminology and the former is the new.  I know there are still remnants of code that refer to "fork" instead of "executeOn", but the trend should be toward fixing those instead of introducing new uses of the old name.